lib/processor/geoip.go

```diff
@@ -0,0 +1,171 @@
+package processor
+
+import (
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/Jeffail/benthos/v3/lib/log"
+	"github.com/Jeffail/benthos/v3/lib/metrics"
+	"github.com/Jeffail/benthos/v3/lib/types"
+	"github.com/opentracing/opentracing-go"
+	"github.com/oschwald/geoip2-golang"
+)
+
+//------------------------------------------------------------------------------
+
+func init() {
+	Constructors[TypeGeoIP] = TypeSpec{
+		constructor: NewGeoIP,
+		Description: `
+		// todo some docs
+		`,
+	}
+}
+
+//------------------------------------------------------------------------------
+
+// GeoIPConfig contains configuration fields for the GeoIP processor.
+type GeoIPConfig struct {
+	Parts        []int  `json:"parts" yaml:"parts"`
+	DatabaseType string `json:"database_type" yaml:"database_type"`
+	Database     string `json:"database" yaml:"database"`
+}
+
+// NewGeoIPConfig returns a GeoIPConfig with default values.
+func NewGeoIPConfig() GeoIPConfig {
+	return GeoIPConfig{
+		Parts:        []int{},
+		DatabaseType: "",
+		Database:     "",
+	}
+}
+
+//------------------------------------------------------------------------------
+
+type geoipReader func(db *geoip2.Reader, ip net.IP) (interface{}, error)
+
+func cityReader() geoipReader {
+	return func(db *geoip2.Reader, ip net.IP) (interface{}, error) {
+		res, err := db.City(ip)
+		if err != nil {
+			return nil, err
+		}
+		return res, nil
```

Comment:
Returning a `*City` type might cause issues downstream as it's a struct. If you don't want to map the entire struct into `map[string]interface{}` then I'll forgive the lazier approach of `json.Marshal(res)` into raw bytes and use `part.Set`. Considering we're doing a DB call here I don't think the performance impact is something to worry about.

Same goes for `asnReader`.